botocore/handlers.py

```diff
@@ -140,6 +143,28 @@ def conditionally_calculate_md5(params, **kwargs):
         calculate_md5(params, **kwargs)
 
 
+def validate_bucket_name(params, **kwargs):
+    # The rules for bucket names in the US Standard region allow bucket names
+    # to be as long as 255 characters, and bucket names can contain any
+    # combination of uppercase letters, lowercase letters, numbers, periods
+    # (.), hyphens (-), and underscores (_).
+    if 'Bucket' not in params:
+        return
+    bucket = params['Bucket']
+    if len(bucket) > 255:
+        error_msg = (
+            'Invalid bucket name "%s": Bucket name cannot be '
+            'longer than 255 characters.' % bucket)
+        raise ParamValidationError(report=error_msg)
+    for char in bucket:
```

Comment:
Any reason to not use a regex here instead? Did you benchmark it and find it to be slower than checking against a list of valid characters?

Alternatively it might be better to use a set for VALID_BUCKET_CHARS so that you don't have to iterate over all of the characters each time your determining if the bucket string is valid. Considering this is going to be used for every S3 request that involves a Bucket parameter (almost all), I think this micro-optimization is reasonable.